lib/ttfunk/table/cff/index.rb

```diff
@@ -0,0 +1,143 @@
+module TTFunk
+  class Table
+    class Cff < TTFunk::Table
+      class Index < TTFunk::SubTable
+        include Enumerable
+
+        # number of objects in the index
+        attr_reader :count
+
+        # offset array element size
+        attr_reader :offset_size
+
+        def [](index)
+          data[index] = get(index)
+        end
+
+        def each
+          return to_enum(__method__) unless block_given?
+          count.times { |i| yield self[i] }
+        end
+
+        def encode
+          result = ''
+
+          entries = each_with_object([]).with_index do |(entry, ret), index|
+            new_entry = block_given? ? yield(entry, index) : entry
+            ret << new_entry if new_entry
+          end
+
+          # "An empty INDEX is represented by a count field with a 0 value and
+          # no additional fields. Thus, the total size of an empty INDEX is 2
+          # bytes."
+          result << [entries.size].pack('n')
+          return result if entries.empty?
+
+          offset_size = (Math.log2(entries.size) / 8.0).round + 1
+          result << [offset_size].pack('c')
+          data_offset = 1
+
+          data = ''
+
+          entries.each do |entry|
+            result << encode_offset(data_offset, offset_size)
+            data << entry
+            data_offset += entry.bytesize
+          end
+
+          unless entries.empty?
+            result << encode_offset(data_offset, offset_size)
+          end
+
+          result << data
+        end
+
+        private
+
+        def encode_offset(offset, offset_size)
+          case offset_size
+          when 1
+            [offset].pack('C')
+          when 2
+            [offset].pack('n')
+          when 3
+            [offset].pack('N')[1..-1]
+          when 4
+            [offset].pack('N')
+          end
+        end
+
+        def get(index)
+          if index >= count
+            raise ArgumentError, "index of #{index} is out-of-bounds"
+          end
+
+          start, finish = relative_data_offsets_for(index)
+          @raw_data_array[start...finish]
+        end
+
+        def parse!
+          @count, @offset_size = read(3, 'nc')
+          @raw_offset_array = io.read((count + 1) * offset_size)
+
+          if @raw_offset_array.empty?
+            @raw_data_array = ''
+            @length = 3
+            return
+          end
+
+          _, last_finish = relative_data_offsets_for(count - 1)
+          @raw_data_array = io.read(last_finish)
```

Comment:
I agree it might be a bit easier to parse the whole thing into an array. However I consciously decided to keep the offset and data strings because I didn't want to spend a lot of time and memory constructing a potentially large array that may not get used anywhere. The `Index` can be accessed by the familiar Ruby indexing strategy (i.e. `#[]`)` and is `Enumerable`, meaning it can be turned into a regular Ruby array pretty easily. The calculation of offsets etc is done in internal, private methods which I had hoped would be abstract enough to be accessible to most Ruby programmers.

Reply:
Well, the whole thing is parsed anyway when the table is re-encoded so no time/memory is saved. I don't think memory overhead is significant anyway. But code complexity is significantly higher compared to upfront parsing.

Reply:
Ah, that's a good point.